File: examples/playground/components/MyStateBag.ts
import {Color, File, FontRegistry, Haiku, Image, IOSFonts, Lottie, Palette, SVG, TextStyle} from '@livedesigner/designsystem';
import {Component, method, property} from '@livedesigner/engine';
import {easeInOutExpo} from 'just-curves';

class MyPalette extends Component<Palette> {
  @property hello = Color.rgba(255, 0, 0, 1);
}

export class MyStateBag extends Component {
  @property palette: MyPalette = new MyPalette();

  @property name: string = 'Diez';

  @property copy = `Hello ${this.name}`;

  @property image = new Image({
    file: new File({
      // Try changing this to diez.jpg!
      src: '/assets/images/haiku.jpg',
    }),
    width: 246,
    height: 246,
    scale: 3,
  });

  @property svg = new SVG({
    file: new File({src: '/assets/images/rat.svg.html'}),
  });

  @property lottie = new Lottie({
    file: new File({src: '/assets/lottie/loading-pizza.json'}),
  });

  // Makes Android fonts available on iOS.
  @property fontRegistry = FontRegistry.fromFiles(
    '/assets/fonts/Roboto-Black.ttf',
    '/assets/fonts/Roboto-BlackItalic.ttf',
    '/assets/fonts/Roboto-Bold.ttf',
    '/assets/fonts/Roboto-BoldItalic.ttf',
    '/assets/fonts/Roboto-Italic.ttf',
    '/assets/fonts/Roboto-Light.ttf',
    '/assets/fonts/Roboto-LightItalic.ttf',
    '/assets/fonts/Roboto-Medium.ttf',
    '/assets/fonts/Roboto-MediumItalic.ttf',
    '/assets/fonts/Roboto-Regular.ttf',
    '/assets/fonts/Roboto-Thin.ttf',
    '/assets/fonts/Roboto-ThinItalic.ttf',
  );

  @property textStyle = new TextStyle({
    font: IOSFonts.Helvetica,
    fontSize: 50,
    color: this.palette.hello,
  });

  @property haiku = new Haiku({
    file: new File({src: '/assets/haiku/animator.html'}),
  });

  @method async tap () {
    await this.palette.hello.tween(
      {h: 0.3},
      // Default curve is linear!
      {duration: 5000},
    );
    await this.palette.hello.tween(
      // You can tween multiple values at once!
      {h: 0.9, l: 0.5},
      {
        duration: 5000,
        // You can specify any pure function (number) -> number as a curve!
        curve: easeInOutExpo,
      },
    );
    await this.palette.tween(
      // You can tween an entire component instead of its inner state!
      {hello: Color.rgba(255, 0, 0, 1)},
      {duration: 5000},
    );
  }
}

File: examples/playground/android/diez/src/main/java/org/diez/components/MyStateBag.kt
package org.diez.components

import android.graphics.Color as CoreColor
import org.diez.color.Color
import org.diez.puente.Method
import org.diez.puente.StateBag
import org.diez.color.ColorAdapter
import org.diez.color.QualifiedColor
import org.diez.color.WireColor
import org.diez.file.File
import org.diez.haiku.Haiku
import org.diez.image.Image
import org.diez.image.SVG
import org.diez.lottie.Lottie
import org.diez.typography.FontRegistry
import org.diez.typography.TextStyle

data class MyPalette(@QualifiedColor var hello: Color = ColorAdapter.hsla(WireColor(0F, 1F, 0.5F, 1F)))

data class MyStateBag(
    var palette: MyPalette = MyPalette(),
    var copy: String = "Hello Diez",
    var image: Image = Image(File("assets/images/haiku.jpg"), 246, 246, 3F),
    var fontRegistry: FontRegistry = FontRegistry(arrayOf<File>(
        File("assets/fonts/Roboto-Black.ttf"),
        File("assets/fonts/Roboto-BlackItalic.ttf"),
        File("assets/fonts/Roboto-Bold.ttf"),
        File("assets/fonts/Roboto-BoldItalic.ttf"),
        File("assets/fonts/Roboto-Italic.ttf"),
        File("assets/fonts/Roboto-Light.ttf"),
        File("assets/fonts/Roboto-LightItalic.ttf"),
        File("assets/fonts/Roboto-Medium.ttf"),
        File("assets/fonts/Roboto-MediumItalic.ttf"),
        File("assets/fonts/Roboto-Regular.ttf"),
        File("assets/fonts/Roboto-Thin.ttf"),
        File("assets/fonts/Roboto-ThinItalic.ttf")
    )),
    var textStyle: TextStyle = TextStyle(
        "Helvetica",
        50F,
        ColorAdapter.hsla(WireColor(0F, 1F, 0.5F, 1F))
    ),
    var haiku: Haiku = Haiku("@haiku/taylor-testthang"),
    var svg: SVG = SVG("assets/images/rat.svg"),
    var lottie: Lottie = Lottie(File("assets/lottie/loading-pizza.json"))
) : StateBag {
    override val adapters = listOf(
        colorAdapter
    )

    override val name = "MyStateBag"

    @Transient
    private var listener: Method? = null

    override fun listen(listener: Method) {
        this.listener = listener
    }
}
